Add binary search

Archargelod opened the pull request:
I used linear_search.nim as base, so `checkBinarySearch` template is redundant, but could be useful in future for other implementations. I am also not sure if it is a good idea or not to check if array is sorted at the start of function; for example:
```
from std/algorithm import isSorted, SortOrder

func binarySearch*[T:Ordinal](arr: openArray[T], key: T): Option[Natural] {.raises: [ValueError].} =
  if not arr.isSorted(Ascending):
    raise newException(ValueError, "Array is not sorted or not in ascending order.")
```

SatinWuker reviewed (commented):
The code looks very good! However, I recommend adding a recursive binary search in order to provide another alternative to people who wants to learn binary search.

Archargelod commented:
> I recommend adding a recursive binary search in order to provide another alternative to people who wants to learn binary search. See https://www.geeksforgeeks.org/binary-search/

Thank you for the link! I added recursive version and fixed a little inconsistency: `mid.Natural`->`Natural(mid)`.

SatinWuker reviewed (approved):
Looks great!

ZoomRmc commented on searches/binary_search.nim:
```diff
@@ -0,0 +1,101 @@
+    check binarySearchRecursive(arr, arr.low(), arr.high(), key) == expected
+
+  suite "Binary Search": 
+    test "search an empty array":
+      var arr: array[0, int]
+      checkBinarySearch(arr, 5, none(Natural))
+
+    test "Search in an int array matching with a valid value":
+      var arr = [0, 1, 2, 3, 5, 6]
+      checkBinarySearch(arr, 5, some(Natural(4)))
+
+    test "Search in an int array for a missing value":
+      var arr = [0, 1, 2, 3, 5, 6]
+      checkBinarySearch(arr, 4, none(Natural))
+
+    test "Search in a char array with a matching value":
+      var arr = ['0', '1', '2', '3', '5', 'a']
+      checkBinarySearch(arr, '5', some(Natural(4)))
+
+    test "Search in an array with a valid key at the end":
+      var arr = [0, 1, 2, 3, 5, 6]
+      checkBinarySearch(arr, 6, some(Natural(5)))
+
+    test "Search in an even-length array with a valid key in the middle":
+      var arr = [0, 1, 2, 3, 5, 6]
+      checkBinarySearch(arr, 3, some(Natural(3)))
+
+    test "Search in an odd-length array with a valid key in the middle":
+      var arr = [0, 1, 2, 3, 5]
+      checkBinarySearch(arr, 2, some(Natural(2)))
```
```suggestion
    const
      empty: array[0, int] = []
      arr = [0, 1, 2, 3, 5, 6]
      odd = [0, 1, 2, 3, 5]
      chars = ['0', '1', '2', '3', '5', 'a']

    test "Empty array":
      checkBinarySearch(empty, 5, none(Natural))

    test "Matching value in an int array":
      checkBinarySearch(arr, 5, some(Natural(4)))

    test "Missing value":
      checkBinarySearch(arr, 4, none(Natural))

    test "Matching value in a char array":
      checkBinarySearch(chars, '5', some(Natural(4)))

    test "Matching key at the start":
      checkBinarySearch(arr, 0, some(Natural(0)))

    test "Matching key at the end":
      checkBinarySearch(arr, 6, some(Natural(5)))

    test "Even-length array with a matching key in the middle":
      checkBinarySearch(arr, 3, some(Natural(3)))

    test "Odd-length array with a matching key in the middle":
      checkBinarySearch(odd, 2, some(Natural(2)))
```

dlesnoff replied:
That's a suggestion I should have made at the Linear Search PR.

ZoomRmc commented on searches/binary_search.nim:
```diff
@@ -0,0 +1,101 @@
+      return binarySearchRecursive(arr, mid + 1, right, key)
+
+  return none(Natural)
+
```
```suggestion

func binarySearchRecursive*[T:Ordinal](arr: openArray[T], key: T): Option[Natural] =
  ## Recursive implementation of binary search for an array sorted in ascending order
  binarySearchRecursive(arr, 0, arr.high(), key)

```
It's idiomatic for Nim to use overloads to simplify user-facing API.

ZoomRmc commented on searches/binary_search.nim:
```diff
@@ -0,0 +1,101 @@
+  # `none(Natural)` is returned when both halfs are empty after some iterations.
+  return none(Natural)
+
+func binarySearchRecursive*[T:Ordinal](arr: openArray[T], left, right: int, key: T): Option[Natural] =
```
```suggestion
func binarySearchRecursive[T:Ordinal](arr: openArray[T], left, right: int, key: T): Option[Natural] =
```

*If the suggestion of using an overload for a public func below is accepted.

ZoomRmc commented on searches/binary_search.nim:
```diff
@@ -0,0 +1,101 @@
+
+  template checkBinarySearch[T:Ordinal](arr: openArray[T], key: T, expected: Option[Natural]): untyped =
+    check binarySearchIterative(arr, key) == expected
+    check binarySearchRecursive(arr, arr.low(), arr.high(), key) == expected
```
```suggestion
    check binarySearchRecursive(arr, key) == expected
```
*If the overload above is accepted.

ZoomRmc reviewed (changes requested):
Thanks for your contribution! Please, consider the proposed changes.

dlesnoff commented on searches/binary_search.nim:
```diff
@@ -0,0 +1,99 @@
+  # `none(Natural)` is returned when both halfs are empty after some iterations.
+  none(Natural)
+
+func binarySearchRecursive[T:Ordinal](arr: openArray[T], left, right: int, key: T): Option[Natural] =
```
```suggestion
func binarySearchRecursive[T:Ordinal](arr: openArray[T], left, right: Natural, key: T): Option[Natural] =
```

Archargelod replied:
Natural here really only adds the complexity as `right` could be -1 (when `mid` = 0). Negative indices can be easily handled by the algorithm without complex edge cases.

ZoomRmc replied:
> Natural here really only adds the complexity as `right` could be -1 (when `mid` = 0). I believe, negative indices can be easily handled by the algorithm without complex edge cases.

There's no negative indexes for `openArrays` and they are necessary to be array/seq agnostic. The other options is to use a concept that will embrace ordinal-indexed types or write type-specific overloads.

dlesnoff commented on searches/binary_search.nim:
```diff
@@ -0,0 +1,99 @@
+## If the key value matches the middle element, the search is complete.
+## If the key value is less than the middle element, the search continues on the lower half of the array.
+## If the key value is greater than the middle element, the search continues on the upper half of the array.
+## This process repeats until the middle element matches the key value or the search space is exhausted.
```
Add a *small* note on other approaches like linear search and search in a tree. See: https://en.wikipedia.org/wiki/Binary_search_algorithm#Binary_search_versus_other_schemes

dlesnoff commented on searches/binary_search.nim:
```diff
@@ -0,0 +1,99 @@
+
+func binarySearchRecursive[T:Ordinal](arr: openArray[T], left, right: int, key: T): Option[Natural] =
+  if left <= right:
+    let mid = left + (right - left) div 2
```
Forget this if you made the change above

dlesnoff commented on searches/binary_search.nim:
```diff
@@ -0,0 +1,133 @@
+## It is intended for simple searches where the presence of the key element is all that matters.
+## https://en.wikipedia.org/wiki/Binary_search_algorithm#Duplicate_elements
+##
+## If the input is not already sorted, it may be faster to use a linear search instead.
```
ChatGPT is not confident. Let us be. Here we can.
```suggestion
## If the input is not already sorted, it is faster to use a linear search instead.
```

dlesnoff reviewed (changes requested):
I changed my mind. Let us just refer to Wikipedia for precisions on theoretical aspects, it will ever be more reviewed and better referenced as well.
Simplify the docstring as suggested, and I'll finally accept the changes.

dlesnoff commented on searches/binary_search.nim:
```diff
@@ -0,0 +1,133 @@
+## If the key value is less than the middle element, the search continues on the lower half of the array.
+## If the key value is greater than the middle element, the search continues on the upper half of the array.
+## This process repeats until the middle element matches the key value or the search space is exhausted.
+## https://en.wikipedia.org/wiki/Binary_search_algorithm
```
Move this reference to the end of the docstring.

Archargelod commented on searches/binary_search.nim:
```diff
@@ -0,0 +1,133 @@
+## Space Complexity in iterative approach: O(1)
+## Space Complexity in recursive approach: O(n)
+##
+## The choice of pivot can greatly affect the performance of a binary search algorithm.
+## This implementation uses the middle element of the array as pivot, it works well
+## when the array is evenly distributed and balanced, as it allows
+## for cutting the search in half with each iteration.
+## Other common choices for pivot include the first or last element of the array,
+## a random element, or the median of three values (the first, middle, and last elements).
```
> Is this A.I. generated?
Yes, sorry for that. I'm not that fluent in english, so I cheated just a bit with docs. 
But don't worry, as code is 100% handwritten and not generated.

ZoomRmc replied:
Next time just say "I could use some help with the docs". Someone will surely offer some necessary additions.

dlesnoff replied:
> But don't worry, as code is 100% handwritten and not generated.

I trust you on that. It is not so much that it has been generated by A.I. but that the documentation was *wrong*.
The code quality is good.

dlesnoff commented:
Run nimpretty to remove trailing whitespaces. You can wrap the arguments on one line by adding a line break after the parenthesis instead of the automatically inserted line break by nimpretty.
I will merge it immediately after.

dlesnoff reviewed (approved):
Nice first contribution!
